### PR TITLE
kubelet: get cgroup driver config from CRI

### PR DESCRIPTION
**Pull request by marquiz**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR implements kubelet cgroup driver detection over the CRI API, implemented in multiple patches:
- __cri-api: add rpc for querying runtime configuration__
  This patch adds a new rpc to the runtime service to query CRI runtime
  configuration options. For now, it only contains one field for getting
  the cgroup driver (systemd or cgroupfs) to be used.
- __Add CRI fake runtimes for RuntimeConfig rpc__
  Also update the CRI RuntimeService inteface to include the new
  RuntimeConfig rpc.
- __Add KubeletCgroupDriverFromCRI feature gate__
- __kubelet: initialization of runtime service earlier in the startup__
  This patch refactors the kubelet startup code to initialize the runtime
  service earlier in the startup sequence. We want this to be able to
  query the cgroup driver setting from the CRI befure initializing the
  cgroup manager.
- __kubelet: get cgroup driver config from CRI__
  This patch modifies kubelet to get the cgroup driver setting from the
  CRI runtime using the newly added RuntimeConfig rpc. The new code path
  only takes place if the KubeletCgroupDriverFromCRI feature gate is
  enabled. If the runtime returns a not-implemented error kubelet falls
  back to using the cgroupDriver configuration option, with a log message
  instructing the user to upgrade to w newer container runtime. Other rpc
  errors cause kubelet to exit as is the case if the runtime returns an
  unknown cgroup driver.

#### Which issue(s) this PR fixes:

Refs https://github.com/kubernetes/enhancements/issues/4033

#### Special notes for your reviewer:

Still work in progress: most importantly unit tests are not included.
If so desired, the changes to cri-api could be submitted as a separate PR.

#### Does this PR introduce a user-facing change?

```release-note
With the KubeletCgroupDriverFromCRI feature gate enabled and sufficiently new version of a container
runtime, kubelet automatically detects the cgroup driver config from the container runtime, eliminating
the need to specify the `cgroupDriver` configuration option (or --cgroup-driver` flag) of kubelet.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4033-group-driver-detection-over-cri

```

### CODE REVIEW

**Comment by marquiz** [comment]
This is an early version of the PR, especially targeted  for bike-shedding on the CRI API naming. It also contains a full implementation (sans unit tests) to see how the new CRI rpc would be used

/cc @mrunalp  @mikebrow @haircommander @sftim

**Comment by sftim** [comment]
Changelog suggestion for when this is ready:
```markdown
Added automatic detection of cgroup driver for Linux nodes.

With the `KubeletCgroupDriverFromCRI` feature gate enabled, and sufficiently new version of a container
runtime, kubelet automatically detects the cgroup driver configuration from the container runtime,
eliminating the need to specify the `cgroupDriver` configuration option (nor the deprecated `--cgroup-driver`
command line argument) for the kubelet.
```

I mentioned Linux because Windows nodes don't have cgroups.

File: cmd/kubelet/app/server.go
Comment (sftim) [inline cmd/kubelet/app/server.go:1306]: For beta, maybe we can even record an Event into the API?

File: cmd/kubelet/app/server.go
Comment (sftim) [inline cmd/kubelet/app/server.go:642]: Do we do this:
- per new connection to the socket?
- per kubelet startup?

I'm thinking that if someone restarts the process at the other end of the socket, we should re-query this setting.

File: cmd/kubelet/app/server.go
Comment (marquiz) [reply cmd/kubelet/app/server.go:1306]: Might be good for the user. The proposal now states that for beta *Drop fallback to old behavior. CRI implementations expected to have support*. We can of course change that, or then add an event for alpha already 🧐

File: cmd/kubelet/app/server.go
Comment (marquiz) [reply cmd/kubelet/app/server.go:642]: This happens only at kubelet startup.

Technically we could query it at CRI reconnection but I wouldn't go there.

1. The cgroup driver is really not something that's expected to be changed without taking the node down. Changing it correctly requires restarting all pods, also the static ones. System reboot is a sage method.
2. I don't know what all would break in kubelet (internally) if we changed the cgroup-driver in-flight

We could re-query the cgroup driver at re-connection and exit if the cgroup driver setting has changed. But I'm not sure if that's desired. Thoughts @SergeyKanzhelev @mrunalp @haircommander

Comment (haircommander) [reply cmd/kubelet/app/server.go:642]: I think our efforts would be better spent making it very clear a change in the cgroup driver in the CRI requires a node reboot

File: cmd/kubelet/app/server.go
Comment (haircommander) [inline cmd/kubelet/app/server.go:633]: nit: we can compress these lines

File: cmd/kubelet/app/server.go
Comment (haircommander) [inline cmd/kubelet/app/server.go:1307]: can there be a comment saying where the cgroup driver is coming from/set by if this case happens

File: cmd/kubelet/app/server.go
Comment (haircommander) [inline cmd/kubelet/app/server.go:1316]: I don't know what the kubelet's policy on panics are, but this seems like a prime reason to panic as it's a programming error that should never happen. I'm not stubborn about that though

Comment (marquiz) [reply cmd/kubelet/app/server.go:1316]: Yeah, It's a programming, but on the CRI runtime's side, out of our (kubelet's) control so I'm not sure we want to panic.

File: cmd/kubelet/app/server.go
Comment (marquiz) [reply cmd/kubelet/app/server.go:1307]: It now states *Falling back to using cgroupDriver from kubelet config*. Do you suggest something else/in addition? @haircommander

**Comment by marquiz** [comment]
@haircommander @mrunalp @mikebrow @sftim any thoughts on the CRI API naming?

The main addition is now 

```protobuf
rpc RuntimeConfig(RuntimeConfigRequest) returns (RuntimeConfigResponse)
```

We have `rpc Status(...)` so maybe it could be just `rpc Configuration(...)` or smth.

But then we have also

```protobuf
rpc UpdateRuntimeConfig(UpdateRuntimeConfigRequest) returns (UpdateRuntimeConfigResponse) {}

message UpdateRuntimeConfigRequest {
    RuntimeConfig runtime_config = 1;
}
```

So the naming is easily clashing...

**Comment by haircommander** [comment]
yeah I guess `RuntimeConfig` is the best we've got so far, considering we use `RuntimeStatus`, `RuntimeCondition`, and `UpdateRuntimeConfig`. 

An issue though is this runtime config has nothing to do with UpdateRuntimeConfig :upside_down_face:

**Comment by mikebrow** [comment]
> @haircommander @mrunalp @mikebrow @sftim any thoughts on the CRI API naming?
> 
> The main addition is now
> 
> ```proto
> rpc RuntimeConfig(RuntimeConfigRequest) returns (RuntimeConfigResponse)
> ```
> 
> We have `rpc Status(...)` so maybe it could be just `rpc Configuration(...)` or smth.
> 
> But then we have also
> 
> ```proto
> rpc UpdateRuntimeConfig(UpdateRuntimeConfigRequest) returns (UpdateRuntimeConfigResponse) {}
> 
> message UpdateRuntimeConfigRequest {
>     RuntimeConfig runtime_config = 1;
> }
> ```
> 
> So the naming is easily clashing...

naming is hard... over on the status api we currently have the ability to provide any desired runtime config response contents ... at least for prototyping just need to agree on the nvp graph/map esp. with status verbose content...

see crictl "info" response

```
(base) root@ubnt:~# crictl info
{
  "status": {
    "conditions": [
      {
        "type": "RuntimeReady",
        "status": true,
        "reason": "",
        "message": ""
      },
      {
        "type": "NetworkReady",
        "status": true,
        "reason": "",
        "message": ""
      }
    ]
  },
  "cniconfig": {
    "PluginDirs": [
      "/opt/cni/bin"
    ],
    "PluginConfDir": "/etc/cni/net.d",
    "PluginMaxConfNum": 1,
    "Prefix": "eth",
    "Networks": [
      {
        "Config": {
          "Name": "cni-loopback",
          "CNIVersion": "0.3.1",
          "Plugins": [
            {
              "Network": {
                "type": "loopback",
                "ipam": {},
                "dns": {}
              },
              "Source": "{\"type\":\"loopback\"}"
            }
          ],
          "Source": "{\n\"cniVersion\": \"0.3.1\",\n\"name\": \"cni-loopback\",\n\"plugins\": [{\n  \"type\": \"loopback\"\n}]\n}"
        },
        "IFName": "lo"
      },
      {
        "Config": {
          "Name": "containerd-net",
          "CNIVersion": "1.0.0",
          "Plugins": [
            {
              "Network": {
                "type": "bridge",
                "ipam": {
                  "type": "host-local"
                },
                "dns": {}
              },
              "Source": "{\"bridge\":\"cni0\",\"ipMasq\":true,\"ipam\":{\"ranges\":[[{\"subnet\":\"10.88.0.0/16\"}],[{\"subnet\":\"2001:4860:4860::/64\"}]],\"routes\":[{\"dst\":\"0.0.0.0/0\"},{\"dst\":\"::/0\"}],\"type\":\"host-local\"},\"isGateway\":true,\"promiscMode\":true,\"type\":\"bridge\"}"
            },
            {
              "Network": {
                "type": "portmap",
                "capabilities": {
                  "portMappings": true
                },
                "ipam": {},
                "dns": {}
              },
              "Source": "{\"capabilities\":{\"portMappings\":true},\"type\":\"portmap\"}"
            }
          ],
          "Source": "{\n  \"cniVersion\": \"1.0.0\",\n  \"name\": \"containerd-net\",\n  \"plugins\": [\n    {\n      \"type\": \"bridge\",\n      \"bridge\": \"cni0\",\n      \"isGateway\": true,\n      \"ipMasq\": true,\n      \"promiscMode\": true,\n      \"ipam\": {\n        \"type\": \"host-local\",\n        \"ranges\": [\n          [{\n            \"subnet\": \"10.88.0.0/16\"\n          }],\n          [{\n            \"subnet\": \"2001:4860:4860::/64\"\n          }]\n        ],\n        \"routes\": [\n          { \"dst\": \"0.0.0.0/0\" },\n          { \"dst\": \"::/0\" }\n        ]\n      }\n    },\n    {\n      \"type\": \"portmap\",\n      \"capabilities\": {\"portMappings\": true}\n    }\n  ]\n}\n"
        },
        "IFName": "eth0"
      }
    ]
  },
  "config": {
    "containerd": {
      "snapshotter": "overlayfs",
      "defaultRuntimeName": "runc",
      "defaultRuntime": {
        "runtimeType": "",
        "runtimePath": "",
        "runtimeEngine": "",
        "PodAnnotations": [],
        "ContainerAnnotations": [],
        "runtimeRoot": "",
        "options": {},
        "privileged_without_host_devices": false,
        "baseRuntimeSpec": "",
        "cniConfDir": "",
        "cniMaxConfNum": 0
      },
      "untrustedWorkloadRuntime": {
        "runtimeType": "",
        "runtimePath": "",
        "runtimeEngine": "",
        "PodAnnotations": [],
        "ContainerAnnotations": [],
        "runtimeRoot": "",
        "options": {},
        "privileged_without_host_devices": false,
        "baseRuntimeSpec": "",
        "cniConfDir": "",
        "cniMaxConfNum": 0
      },
      "runtimes": {
        "runc": {
          "runtimeType": "io.containerd.runc.v2",
          "runtimePath": "",
          "runtimeEngine": "",
          "PodAnnotations": [],
          "ContainerAnnotations": [],
          "runtimeRoot": "",
          "options": {
            "BinaryName": "",
            "CriuImagePath": "",
            "CriuPath": "",
            "CriuWorkPath": "",
            "IoGid": 0,
            "IoUid": 0,
            "NoNewKeyring": false,
            "NoPivotRoot": false,
            "Root": "",
            "ShimCgroup": "",
            "SystemdCgroup": false
          },
          "privileged_without_host_devices": false,
          "baseRuntimeSpec": "",
          "cniConfDir": "",
          "cniMaxConfNum": 0
        }
      },
      "noPivot": false,
      "disableSnapshotAnnotations": true,
      "discardUnpackedLayers": false,
      "ignoreRdtNotEnabledErrors": false
    },
    "cni": {
      "binDir": "/opt/cni/bin",
      "confDir": "/etc/cni/net.d",
      "maxConfNum": 1,
      "confTemplate": "",
      "ipPref": ""
    },
    "registry": {
      "configPath": "",
      "mirrors": {},
      "configs": {},
      "auths": {},
      "headers": {}
    },
    "imageDecryption": {
      "keyModel": "node"
    },
    "disableTCPService": true,
    "streamServerAddress": "127.0.0.1",
    "streamServerPort": "0",
    "streamIdleTimeout": "4h0m0s",
    "enableSelinux": false,
    "selinuxCategoryRange": 1024,
    "sandboxImage": "registry.k8s.io/pause:3.8",
    "statsCollectPeriod": 10,
    "systemdCgroup": false,
    "enableTLSStreaming": false,
    "x509KeyPairStreaming": {
      "tlsCertFile": "",
      "tlsKeyFile": ""
    },
    "maxContainerLogSize": 16384,
    "disableCgroup": false,
    "disableApparmor": false,
    "restrictOOMScoreAdj": false,
    "maxConcurrentDownloads": 3,
    "disableProcMount": false,
    "unsetSeccompProfile": "",
    "tolerateMissingHugetlbController": true,
    "disableHugetlbController": true,
    "device_ownership_from_security_context": false,
    "ignoreImageDefinedVolumes": false,
    "netnsMountsUnderStateDir": false,
    "enableUnprivilegedPorts": false,
    "enableUnprivilegedICMP": false,
    "containerdRootDir": "/var/lib/containerd",
    "containerdEndpoint": "/run/containerd/containerd.sock",
    "rootDir": "/var/lib/containerd/io.containerd.grpc.v1.cri",
    "stateDir": "/run/containerd/io.containerd.grpc.v1.cri"
  },
  "golang": "go1.19.9",
  "lastCNILoadStatus": "OK",
  "lastCNILoadStatus.default": "OK"
}

```

**Comment by mikebrow** [comment]
> yeah I guess `RuntimeConfig` is the best we've got so far, considering we use `RuntimeStatus`, `RuntimeCondition`, and `UpdateRuntimeConfig`.
> 
> An issue though is this runtime config has nothing to do with UpdateRuntimeConfig 🙃

The original idea for UpdateRuntimeConfig was to add more optional structs as we added more runtime configuration options for kubelet..

**Comment by haircommander** [comment]
> The original idea for UpdateRuntimeConfig was to add more optional structs as we added more runtime configuration options for kubelet..

not within scope here: something we should think about long term is the direction of data. UpdateRuntimeConfig tells cri to update (kubelet->cri). RuntimeConfig is kubelet asking CRI what to do (cri->kubelet). as time goes on, we probably want to choose one direction of information

**Comment by mikebrow** [comment]
after discussion with @haircommander perhaps the new runtime service cri api should be StaticRuntimeConfig to cover retrieving the cgroup driver for now and we can add more content that won't change, without a reboot of the container runtime as needed...

File: cmd/kubelet/app/server.go
Comment (marquiz) [reply cmd/kubelet/app/server.go]: Changed as suggested

**Comment by marquiz** [comment]
> perhaps the new runtime service cri api should be StaticRuntimeConfig to cover retrieving the cgroup driver for now and we can add more content that won't change, without a reboot of the container runtime as needed...

The only aspect I'm thinking is how good are we predicting what will be static going into the future 🧐 What if something here changes to dynamic in nature in future versions of the CRI runtimes? Just add it to the "DynamicConfiguration" rpc as well, even if it's not static anymore and be happy with it(?)

**Comment by haircommander** [comment]
I've added a commit adding additional context about the expectations around cgroup driver and restarts.

I am going to unmark as WIP so we can get reviews flowing. This is about ready besides naming convention consensus

/retitle kubelet: get cgroup driver config from CRI

**Comment by marquiz** [comment]
Thanks for the help @haircommander (while I'm on PTO), The of API notes are a good addition

**Review by mikebrow — APPROVED** [review]
LGTM

**Comment by haircommander** [comment]
@SergeyKanzhelev @mrunalp @sftim PTAL

File: cmd/kubelet/app/server.go
Comment (SergeyKanzhelev) [inline cmd/kubelet/app/server.go:1309]: should there be `nil` checks?

File: cmd/kubelet/app/server.go
Comment (SergeyKanzhelev) [reply cmd/kubelet/app/server.go:1315]: let's fall back to the configuration file for now

Comment (SergeyKanzhelev) [reply cmd/kubelet/app/server.go:1315]: Thinking more... this should be treated differently than if runtime is not available, for example, or returned error, going forward. I think if runtime returned error we will need to retry a few times and fallback to config. If it's a new unknown value, kubelet should not retry and should exit immediately.

File: cmd/kubelet/app/server.go
Comment (haircommander) [reply cmd/kubelet/app/server.go]: updated as suggested!

**Comment by haircommander** [comment]
@SergeyKanzhelev @mrunalp @sftim PTAL

**Comment by haircommander** [comment]
@SergeyKanzhelev @mrunalp @sftim PTAL

File: cmd/kubelet/app/server.go
Comment (mrunalp) [inline cmd/kubelet/app/server.go:1305]: I am not sure that adding a loop here is adding much value. What transient errors do we expect to succeed on retries? Also, there is no backoff or wait in between attempts.

File: cmd/kubelet/app/server.go
Comment (haircommander) [reply cmd/kubelet/app/server.go:1334]: @mrunalp I added it bc of @SergeyKanzhelev 's comment. I am personally fine without it, i'll let y'all decide :)

File: cmd/kubelet/app/server.go
Comment (SergeyKanzhelev) [reply cmd/kubelet/app/server.go:1305]: if kubelet and runtime are started together, transient issue may be just a timing of a start. Before we were establishing connection a bit later in a function - after reading some metrics and such. We can exit kubelet immediately and rely on external restart of `kubelet`. Or we can do a few iterations here. But if we do retry here - there should be a delay between attempts

Comment (mrunalp) [reply cmd/kubelet/app/server.go:1305]: Yes, without delay there isn't much value here.

Comment (haircommander) [reply cmd/kubelet/app/server.go:1305]: how is 2 seconds? I had trouble finding a value to base it on so I just made a guess.

Comment (marquiz) [reply cmd/kubelet/app/server.go:1304]: I'm back from PTO, thanks for all the help with this @kubernetes/milestone-maintainers 

How long would we expect for the CRI runtime startup to take? Just wonder should the sleep be 1 or 2 seconds or smth else.

**Review by mrunalp — APPROVED** [review]

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

A few notes:

1. The call to `PreInitRuntimeService` was moved to earlier in the flow unconditionally. This change the behavior when feature gate is not set. Looking at what is being done between the new callsite and the old callsite - the risk that it will break something is very low.
2. Changelog file for CRI API will need to be updated. Close to code freeze it's best to do as a follow up as there is another PR updating it now.

**Comment by saschagrunert** [comment]
@marquiz is the release note still **_WIP_** or can we remove that prefix?

**Comment by marquiz** [comment]
> @marquiz is the release note still **_WIP_** or can we remove that prefix?

I removed the wip prefix

**Comment by saschagrunert** [comment]
Thank you @marquiz! :pray:

**Pull request merged**
